pkg/admission/mutator/shoot_mutator.go

```diff
@@ -0,0 +1,288 @@
+	}
+
+	versions := falco.FalcoVersions().Falco
+	version, err := chooseHighestVersion(versions, "supported")
+	if err != nil {
+		return err
+	}
+	falcoConf.FalcoVersion = version
+	return nil
+}
+
+func chooseHighestVersion(versions *falcoversions.FalcoVersions, classification string) (*string, error) {
+	var highest *pkgversion.Version = nil
+	for _, v := range versions.FalcoVersions {
+		if v.Classification != classification {
+			continue
+		}
+		incumbent, err := pkgversion.NewVersion(v.Version)
+
+		if err != nil {
+			return nil, fmt.Errorf("could not parse version: %s", err)
+		}
+
+		if highest == nil || incumbent.GreaterThan(highest) {
+			highest = incumbent
+		}
+	}
+
+	if highest != nil {
+		highestVersion := highest.String()
+		return &highestVersion, nil
+	}
+	return nil, fmt.Errorf("no version with classification %s was found", classification)
+}
+
+func (s *shoot) mutateShoot(_ context.Context, new *gardencorev1beta1.Shoot) error {
+	if s.isDisabled(new) {
+		return nil
+	}
+	falcoConf, err := s.extractFalcoConfig(new)
+	if err != nil {
+		return err
+	}
+
+	if err = setFalcoVersion(falcoConf); err != nil {
+		return err
+	}
+
+	setAutoUpdate(falcoConf)
+
+	setResources(falcoConf)
+
+	setFalcoCtl(falcoConf)
+
+	setGardenerRules(falcoConf)
+
+	setCustomWebhook(falcoConf)
+
+	return s.updateFalcoConfig(new, falcoConf)
+
+	// oldNamedResources := map[string]int{}
+	// for i, r := range new.Spec.Resources {
+	// 	oldNamedResources[r.Name] = i
+	// }
+	// newNamedResources := map[string]struct{}{}
+
+	// outdated := map[string]struct{}{}
+	// for key := range oldNamedResources {
+	// 	if !strings.HasPrefix(key, pkgservice.ExtensionType+"-") {
+	// 		continue
+	// 	}
+	// 	if _, ok := newNamedResources[key]; !ok {
+	// 		outdated[key] = struct{}{}
+	// 	}
+	// }
+	// if len(outdated) > 0 {
+	// 	newResources := []gardencorev1beta1.NamedResourceReference{}
+	// 	for _, resource := range new.Spec.Resources {
+	// 		if _, ok := outdated[resource.Name]; !ok {
+	// 			newResources = append(newResources, resource)
+	// 		}
+	// 	}
+	// 	new.Spec.Resources = newResources
+	// }
+
+	// return s.updateDNSConfig(new, dnsConfig)
+}
+
+// isDisabled returns true if extension is explicitly disabled.
+func (s *shoot) isDisabled(shoot *gardencorev1beta1.Shoot) bool {
+	if shoot.DeletionTimestamp != nil {
+		// don't mutate shoots in deletion
+		return true
+	}
+
+	ext := s.findExtension(shoot)
+	if ext == nil {
+		return true
+	}
+	if ext.Disabled != nil {
+		return *ext.Disabled
+	}
+	return false
+}
+
+// findExtension returns shoot-falco-service extension.
+func (s *shoot) findExtension(shoot *gardencorev1beta1.Shoot) *gardencorev1beta1.Extension {
+	extensionType := "shoot-falco-service"
+	for i, ext := range shoot.Spec.Extensions {
+		if ext.Type == extensionType {
+			return &shoot.Spec.Extensions[i]
+		}
+	}
+	return nil
+}
+
+func (s *shoot) extractFalcoConfig(shoot *gardencorev1beta1.Shoot) (*service.FalcoServiceConfig, error) {
```

Comment:
can we move this to a utils class? Similar coding is used in other places as well.

Reply:
Needs to stay as shoot is different struct for validator and mutator